config/manager/manager.yaml

```diff
@@ -11,6 +11,36 @@ spec:
   ports:
     - port: 8443
 ---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: addon-manager-configmap
+  namespace: addon-manager-system
+data:
+  addon-manager-controller.yaml: |
+    addon-namespaces:
```

Comment:
What's this?

Reply:
This is the list of namespaces that addon-manager is going to list/watch workflow custom resources.

Reply:
We should not expose corp use-case to opensource users, they will not have these addons. Remove these.

Reply:
Also, why would Addon-Manager need to be pre told what namespaces are being watched, the Addon `.spec.params.namespace` is meant to be the target namespace where resources would be deployed but it might not always be the case.

We would better to have some label or annotation applied to namespace to enable watches, it shouldn't be static like this.

Reply:
First of all, addon-manager is not allowed to list/watch all namespaces due to because of role definition
https://github.com/keikoproj/addon-manager/blob/master/config/rbac/role.yaml#L105

Secondly, the controller-runtime/manager framework's rule https://github.com/keikoproj/addon-manager/pull/126/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261L60
https://github.com/keikoproj/addon-manager/pull/126/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R66
allows user to specify multiple namespaces through controller/options only. 
User is not allowed to dynamically tell controller-runtime/controller list/watch namespaces

hope this detail can clarify your doubt.

Reply:
1. There's a cluster-role that allows rbac to namespaces and other default resources https://github.com/keikoproj/addon-manager/blob/612ebb4d50bc8abb19392b5edcde18b8d7150e89/config/rbac/role.yaml#L4-L15
2. We have done that in master, we setup global watches and also dynmaically setup [observeResource](https://github.com/keikoproj/addon-manager/blob/612ebb4d50bc8abb19392b5edcde18b8d7150e89/controllers/addon_controller.go#L549)

We do not want this way, it's not make sense.

Reply:
we have a well-known issue of cluster-wide namespace watch
https://github.com/keikoproj/addon-manager/issues/43

Reply:
That was for Workflow resources, generally we can watch lower threshold resources like `Deployments` we shouldn't have 10K which is possible w/ Workflows.

Reply:
But, even with limited workflows, when watch from cluster-wide namespace, we have serious resource consumptio issue, which is fixed https://github.com/keikoproj/addon-manager/pull/73 here
We do not want to re-introduce those issue again.

Reply:
Yea, but we currently don't expect to have addon namespaces pre-created, this design is a problem. The Addon manager should not be aware of addon resources until the Addon spec is submitted, that includes namespaces. Let's remove this code it's unwanted feature for now.

Reply:
@kevdowney  Those namespaces are not pre-created, when we configure one addon, we specify the addon namespace when it is configured for deployment, here is the an example https://github.intuit.com/kubernetes/arktika/blob/master/config/addons/cron-pdb-reaper/addon.yaml#L17
This tiny, small solution is  just use existing data resolve the problem addon-manager is facing.

Yes, you are definitely right, addon-manager is only aware of the spec when configuration is done. So, we pick configmap which can be configured during deployment and loaded into addon-manager after the addon-spec is submitted.

Reply:
Put it simply, for the release, or, for a specific cluster, IKS.yaml defines all externalAddons/clusterAddons, as well as the addons' namespaces, just make the namespaces consistent with the addon-manager-configmap configuration. So that addon-manager will load the data https://github.com/keikoproj/addon-manager/pull/126/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R52 dynamically once the IKS is deployed by arktika

Reply:
7ccee9f3c8ed3d1056b7c3d44d26c0a9faf141fc
removed